### PR TITLE
[Refactor]: Unified parameter initialization

### PR DESCRIPTION
**Pull request by xiliu8006**
## Motivation
Refactoring model initialization methods based MMCV. For more information, please refer to open-mmlab/mmcv#780

## Dependency
- [x] open-mmlab/mmcv#780
- [x] open-mmlab/mmcv#884
- [x] open-mmlab/mmcv#893
- [x] open-mmlab/mmcv#899
- [x] open-mmlab/mmcv#902
- [x] open-mmlab/mmcv#918
- [x] open-mmlab/mmdetection#4750

## Training results
  | Methods | config | Before refactor init_weight() | After  after init_weight() |
  |:-------:|:-----------:|:-------------:|:---------:|
  |3DSSD| [config](https://github.com/open-mmlab/mmdetection3d/blob/b255d16b9f8a19a8f1065a1ae6c88fd2a03dc325/configs/3dssd/3dssd_kitti-3d-car.py) | 78.39(mAP)| 78.30(mAP)|
  | VoteNet | [config](https://github.com/open-mmlab/mmdetection3d/blob/b255d16b9f8a19a8f1065a1ae6c88fd2a03dc325/configs/votenet/votenet_16x8_sunrgbd-3d-10class.py)        |59.07(AP@0.25), 35.77(AP@0.5)|  60.01(AP@0.25), 36.77(AP@0.5)|         |
  | PointPillars-car|  [config](https://github.com/open-mmlab/mmdetection3d/blob/b255d16b9f8a19a8f1065a1ae6c88fd2a03dc325/configs/pointpillars/hv_pointpillars_secfpn_6x8_160e_kitti-3d-car.py)        | 77.1(mAP)| 77.12(mAP)         |
  | PartA2| [config](https://github.com/open-mmlab/mmdetection3d/blob/b255d16b9f8a19a8f1065a1ae6c88fd2a03dc325/configs/parta2/hv_PartA2_secfpn_2x8_cyclic_80e_kitti-3d-car.py)|   79.16(mAP)    |  79.47(mAP)   |
  | SECOND| [config](https://github.com/open-mmlab/mmdetection3d/blob/951b521985e9174b392bb463f2f38d098cf1c8c5/configs/second/hv_second_secfpn_6x8_80e_kitti-3d-car.py) | 79.07(mAP)           |78.73(mAP)     | 
  | imVoteNet| [config](https://github.com/open-mmlab/mmdetection3d/blob/951b521985e9174b392bb463f2f38d098cf1c8c5/configs/imvotenet/imvotenet_stage2_16x8_sunrgbd-3d-10class.py) |  64.04(AP@0.25)        | 63.82 (AP@0.25)     | 
  | CenterPoint|[config](https://github.com/open-mmlab/mmdetection3d/blob/86ef23d437f4d919412d41219b3b2fbe6c596fbe/configs/centerpoint/centerpoint_01voxel_second_secfpn_circlenms_4x8_cyclic_20e_nus.py)| 56.19(mAP), 64.43(NDS)| 0.5627(mAP), 0.6459(NDS)|
  |MVX-Net|[config](https://github.com/open-mmlab/mmdetection3d/blob/951b521985e9174b392bb463f2f38d098cf1c8c5/configs/mvxnet/dv_mvx-fpn_second_secfpn_adamw_2x8_80e_kitti-3d-3class.py)| 63.0(mAP)| 62.50(mAP)|
  |Regnet| [config](https://github.com/open-mmlab/mmdetection3d/blob/86ef23d437f4d919412d41219b3b2fbe6c596fbe/configs/regnet/hv_pointpillars_regnet-400mf_secfpn_sbn-all_4x8_2x_nus-3d.py)| 41.2(mAP), 55.2(NDS)| 41.4(mAP), 54.3(NDS), batch_size=1| 
  |H3dnet|[config](https://github.com/open-mmlab/mmdetection3d/blob/b255d16b9f8a19a8f1065a1ae6c88fd2a03dc325/configs/h3dnet/h3dnet_3x8_scannet-3d-18class.py)| 66.43(AP@0.25), 48.01(AP@0.5) |65.76(AP@0.25), 47.32(AP@0.5)|
  |SSN| [config](https://github.com/open-mmlab/mmdetection3d/blob/43d79534d9fbbc4bcfe8bcdecca02644c9844f27/configs/ssn/hv_ssn_regnet-400mf_secfpn_sbn-all_2x16_2x_nus-3d.py)| 41.56(mAP), 54.83(NDS)|39.38(mAP), 54.01(NDS)|
 |FCOS3D|[config]()| 29.9(mAP), 37.3(NDS)|26.53(mAP), 34.07(NDS)|

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Need to resolve conflicts. FCOS3D seems to have low accuracy.

File: mmdet3d/models/dense_heads/centerpoint_head.py [hunk 1]
```diff
@@ -78,15 +80,18 @@ def __init__(self,
 
             self.__setattr__(head, conv_layers)
 
+            if init_cfg is None:
+                self.init_cfg = dict(
+                    type='Kaiming',
+                    layer='Conv2d',
+                    override=dict(type='Kaiming', name='heatmap'))
```
Comment (MeowZheng) [inline mmdet3d/models/dense_heads/centerpoint_head.py:87]: I think override can be deleted, as you have init ``heatmap`` in init_weights.

File: mmdet3d/models/dense_heads/centerpoint_head.py [hunk 2]
```diff
@@ -192,11 +200,14 @@ def __init__(self,
...
-        """Initialize weights."""
-        self.cls_head[-1].bias.data.fill_(self.init_bias)
-        self.task_head.init_weights()
+        if init_cfg is None:
+            self.init_cfg = dict(
+                type='Kaiming',
+                layer='Conv2d',
+                override=dict(
+                    type='Kaiming',
+                    name=self.cls_head[-1],
```
Comment (MeowZheng) [inline mmdet3d/models/dense_heads/centerpoint_head.py:209]: the value of ``name`` is attribute name. I suggest to rewrite ``init_weights`` for initializing cls_head.

**Review by MeowZheng — COMMENTED** [review]
After [pr](https://github.com/open-mmlab/mmcv/pull/1150) merging, please check the log carefully.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -91,7 +91,7 @@ jobs:
       - name: Install mmdet3d dependencies
         run: |
           pip install mmcv-full -f https://download.openmmlab.com/mmcv/dist/cu101/${{matrix.torch_version}}/index.html
-          pip install mmdet==2.11.0
+          pip install -q git+https://github.com/open-mmlab/mmdetection.git
```
Comment (ZwwWayne) [inline .github/workflows/build.yml:94]: pip install mmdet==2.14.0
pip install mmseg==0.14.1

File: mmdet3d/__init__.py [hunk 4]
```diff
@@ -17,7 +17,7 @@ def digit_version(version_str):
     return digit_version
 
 
-mmcv_minimum_version = '1.3.1'
+mmcv_minimum_version = '1.3.5'
```
Comment (ZwwWayne) [inline mmdet3d/__init__.py:20]: Should set 1.3.8

File: mmdet3d/__init__.py [hunk 5]
```diff
@@ -27,17 +27,17 @@ def digit_version(version_str):
     f'MMCV=={mmcv.__version__} is used but incompatible. ' \
     f'Please install mmcv>={mmcv_minimum_version}, <={mmcv_maximum_version}.'
 
-mmdet_minimum_version = '2.10.0'
-mmdet_maximum_version = '2.11.0'
+mmdet_minimum_version = '2.12.0'
```
Comment (ZwwWayne) [inline mmdet3d/__init__.py:30]: Should be 2.14.0 then.

File: mmdet3d/__init__.py [hunk 6]
```diff
@@ -27,17 +27,17 @@ def digit_version(version_str):
...
 assert (mmdet_version >= digit_version(mmdet_minimum_version)
         and mmdet_version <= digit_version(mmdet_maximum_version)), \
     f'MMDET=={mmdet.__version__} is used but incompatible. ' \
     f'Please install mmdet>={mmdet_minimum_version}, ' \
     f'<={mmdet_maximum_version}.'
 
-mmseg_minimum_version = '0.14.0'
-mmseg_maximum_version = '0.14.0'
+mmseg_minimum_version = '0.14.1'
+mmseg_maximum_version = '0.15.0'
```
Comment (ZwwWayne) [inline mmdet3d/__init__.py:40]: may set 1.0.0 for now. Usually there is not many BC-breakings.

File: docs/compatibility.md [hunk 7]
```diff
@@ -4,6 +4,15 @@ This document provides detailed descriptions of the BC-breaking changes in MMDet
 
 ## MMDetection3D 0.15.0
 
+### Unified parameter initialization
+
+To unify the parameter initialization in OpenMMLab projects, MMCV supports `BaseModule` that accepts `init_cfg` to allow the modules' parameters initialized in a flexible and unified manner. Now the users need to explicitly call `model.init_weights()` in the training script to initialize the model (as in [here](https://github.com/open-mmlab/mmdetection3d/blob/master/tools/train.py#L183), previously this was handled by the detector. Please refer to PR #378 for details.
```
Comment (ZwwWayne) [inline docs/compatibility.md:9]: change #378 to #622

File: docs/compatibility.md [hunk 8]
```diff
@@ -4,6 +4,15 @@ This document provides detailed descriptions of the BC-breaking changes in MMDet
...
+
+### BackgroundPointsFilter
+
+We modified the dataset aumentation function `BackgroundPointsFilter`(in [here](https://github.com/open-mmlab/mmdetection3d/blob/mmdet3d/datasets/pipelines/transforms_3d.py#L1101)). In previous version of MMdetection3D, `BackgroundPointsFilter` changes the gt_bboxes_3d's bottom center to the gravity center. In MMDetection3D 0.14.0,
```
Comment (ZwwWayne) [inline docs/compatibility.md:13]: 0.14.0?

File: mmdet3d/apis/inference.py [hunk 9]
```diff
@@ -66,6 +66,7 @@ def init_model(config, checkpoint=None, device='cuda:0'):
     model.cfg = config  # save the config in the model for convenience
     model.to(device)
     model.eval()
+    model.init_weights()
```
Comment (ZwwWayne) [inline mmdet3d/apis/inference.py:69]: We do not need to initialize the model because it loads the checkpoint. If we initialize the model the checkpoint will have no effect.

File: tests/test_data/test_pipelines/test_augmentations/test_transforms_3d.py [hunk 10]
```diff
@@ -623,6 +624,8 @@ def test_background_points_filter():
     assert repr_str == expected_repr_str
     assert points.shape == (800, 4)
     assert np.allclose(orig_points, points)
+    assert np.allclose(input_dict['gt_bboxes_3d'].tensor.numpy(),
```
Comment (ZwwWayne) [inline tests/test_data/test_pipelines/test_augmentations/test_transforms_3d.py:627]: This modification seems to be irrelavent.

File: mmdet3d/models/backbones/second.py [hunk 11]
```diff
@@ -61,14 +64,14 @@ def __init__(self,
...
+            'init_cfg and pretrained cannot be setting at the same time'
         if isinstance(pretrained, str):
-            from mmdet3d.utils import get_root_logger
-            logger = get_root_logger()
-            load_checkpoint(self, pretrained, strict=False, logger=logger)
+            warnings.warn('DeprecationWarning: pretrained is a deprecated, '
+                          'please use "init_cfg" instead')
+            self.init_cfg = dict(type='Pretrained', checkpoint=pretrained)
+        else:
+            self.init_cfg = dict(type='Kaiming', layer='Conv2d')
```
[the remark this answers is not in the document]
Comment (xiliu8006) [reply mmdet3d/models/backbones/second.py:74]: if we dont use the pretrained model, we want to use kaiming_init to init all  Conv2d layer.

File: mmdet3d/models/dense_heads/centerpoint_head.py (hunk 1 above)
Comment (xiliu8006) [reply mmdet3d/models/dense_heads/centerpoint_head.py:87]: > Yes, I forget to delete it. thank you!

File: docs/compatibility.md (hunk 8 above)
Comment (xiliu8006) [reply docs/compatibility.md:13]: Yes, This PR is merged in MMDetection 0.14.0. The PR #621 Only modifies the unittest
